app/platform/platform.c

```diff
@@ -135,41 +159,62 @@ int platform_gpio_read( unsigned pin )
 }
 
 #ifdef GPIO_INTERRUPT_ENABLE
-static void platform_gpio_intr_dispatcher( void *arg) {
-  platform_gpio_intr_handler_fn_t cb = arg;
-  uint8 i, level;
+static task_handle_t gpio_task_handle;
+
+static void ICACHE_RAM_ATTR platform_gpio_intr_dispatcher (void *dummy){
+  uint32 j=0;
   uint32 gpio_status = GPIO_REG_READ(GPIO_STATUS_ADDRESS);
-  for (i = 0; i < GPIO_PIN_NUM; i++) {
-    if (pin_int_type[i] && (gpio_status & BIT(pin_num[i])) ) {
-      //disable interrupt
-      gpio_pin_intr_state_set(GPIO_ID_PIN(pin_num[i]), GPIO_PIN_INTR_DISABLE);
-      //clear interrupt status
-      GPIO_REG_WRITE(GPIO_STATUS_W1TC_ADDRESS, gpio_status & BIT(pin_num[i]));
-      level = 0x1 & GPIO_INPUT_GET(GPIO_ID_PIN(pin_num[i]));
-      if(cb){
-        cb(i, level);
+  UNUSED(dummy);
+  /*
+   * gpio_status is a bit map where bit 0 is set if unmapped gpio pin 0 (pin3) has 
+   * triggered the ISR. bit 1 if unmapped gpio pin 1 (pin10=U0TXD), etc.  Since this
+   * in the ISR, it makes sense to optimize this by doing a fast scan of the status
+   * and reverse mapping any set bits.
+   */
+   for (j = 0; gpio_status>0; j++, gpio_status >>= 1) {
+    if (gpio_status&1) {
+      int i = pin_num_inv[j];
+      if (pin_int_type[i]) {
+        //disable interrupt
+        gpio_pin_intr_state_set(GPIO_ID_PIN(j), GPIO_PIN_INTR_DISABLE);
+        //clear interrupt status
+        GPIO_REG_WRITE(GPIO_STATUS_W1TC_ADDRESS, BIT(j));
+        uint32 level = 0x1 & GPIO_INPUT_GET(GPIO_ID_PIN(j));
+        if (pin_trigger[i]) {
+          /* the task is only posted if a trigger callback is defined */
+          pin_trigger[i] = false;
+          task_post_high (gpio_task_handle, (i<<1) + level);
+        }
+       // Interrupts are re-enabled but any interrupt occuring before pin_trigger[i] is reset will be ignored.
+      gpio_pin_intr_state_set(GPIO_ID_PIN(j), pin_int_type[i]);  
       }
-      gpio_pin_intr_state_set(GPIO_ID_PIN(pin_num[i]), pin_int_type[i]);
     }
   }
 }
 
-void platform_gpio_init( platform_gpio_intr_handler_fn_t cb )
+void platform_gpio_init( task_handle_t gpio_task )
 {
-  ETS_GPIO_INTR_ATTACH(platform_gpio_intr_dispatcher, cb);
-}
+  int i;
+  gpio_task_handle = gpio_task;
 
-int platform_gpio_intr_init( unsigned pin, GPIO_INT_TYPE type )
+  get_pin_map();
+  ETS_GPIO_INTR_ATTACH(platform_gpio_intr_dispatcher, NULL);
+}
+/*
+ * Initialise GPIO interrupt mode. In RAM because interrupts are dsabled 
```

Comment:
Is this really the case that code with interrupts masked must be run out of RAM? The i2c code doesn't do this. Given the small amount of space that we have in the float build, I'd like to make sure one way or the other.